gw_eccentricity/eccDefinition.py

```diff
@@ -1285,33 +1442,217 @@ def get_good_extrema(self, pericenters, apocenters,
+                    = self.special_interp_kwargs_for_extrema["denom_degree"] + 1
+                self.special_interp_kwargs_for_extrema["num_degree"],\
+                    self.special_interp_kwargs_for_extrema["denom_degree"] \
+                    = new_num_degree, new_denom_degree
+                # build new fit and check monotonicity
+                new_rat_fit = get_rational_fit(
+                    x, y,
+                    rational_fit_kwargs=self.special_interp_kwargs_for_extrema,
+                    check_kwargs=False)
+                fit_is_nonmonotonic \
+                    = self.check_if_first_derivative_is_not_strictly_monotonic(
+                        x_test, new_rat_fit(x_test), 1.0, data_name)
+                if fit_is_nonmonotonic:
+                    # Revert to previous fit and degrees if nonmonotonicity is
+                    # detected
+                    debug_message(
+                        "Increasing degrees caused nonmonotonicity. "
+                        "Reverting to last monotonic fit with "
+                        f"num_degree={last_monotonic_num_degree} and "
+                        f"denom_degree={last_monotonic_denom_degree}.",
+                        debug_level=self.debug_level, important=False)
+                    self.special_interp_kwargs_for_extrema["num_degree"] \
+                        = last_monotonic_num_degree
+                    self.special_interp_kwargs_for_extrema["denom_degree"] \
+                        = last_monotonic_denom_degree
+                    rat_fit = last_monotonic_rat_fit
+                    break
+
+                last_monotonic_rat_fit = new_rat_fit
+                last_monotonic_num_degree = new_num_degree
+                last_monotonic_denom_degree = new_denom_degree
+                rat_fit = new_rat_fit
+
+        # update final degrees used to build the fit
+        self.rational_fit_degrees[data_name] = (
+            self.special_interp_kwargs_for_extrema["num_degree"],
+            self.special_interp_kwargs_for_extrema["denom_degree"])
+        return rat_fit
+
+    def get_approximate_degree_for_rational_fit(self):
+        """Get approximate degree based on the number of extrema.
+
+        Assign degree based on number of extrema found. The degree is increased
+        as the number of extrema increases.
+        """
+        # TODO: Optimize this. These numbers are not very robust
+        # since the relation between the final degrees and number of
+        # cycles is not very simple, and are meant only for a
+        # reasonable guess for the initial degrees which is then
+        # adapted to get the final degrees.
+        
+        # assign degree based on the number of extrema if user provided
+        # degree is None.
+        approximate_num_orbits = max(len(self.pericenters_location),
+                                     len(self.apocenters_location))
+        if approximate_num_orbits <= 20:
```

Comment:
Maybe make a wiki page with the investigations you did and link that here as motivation for why you use these numbers?

Reply:
Linked a wiki page.